library/asn1write.c

```diff
@@ -234,21 +234,33 @@ int mbedtls_asn1_write_bool( unsigned char **p, unsigned char *start, int boolea
 int mbedtls_asn1_write_int( unsigned char **p, unsigned char *start, int val )
 {
     int ret;
-    size_t len = 0;
+    size_t len;
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+    unsigned i;
+    unsigned char *c;
+#endif
 
-    if( *p - start < 1 )
+    if( *p < start || (size_t)( *p - start ) < sizeof(int) )
         return( MBEDTLS_ERR_ASN1_BUF_TOO_SMALL );
 
-    len += 1;
-    *--(*p) = val;
-
-    if( val > 0 && **p & 0x80 )
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+    c = (unsigned char *)&val;
+    for( i = 0 ; i < sizeof(int) ; i++ )
     {
-        if( *p - start < 1 )
-            return( MBEDTLS_ERR_ASN1_BUF_TOO_SMALL );
+        *--(*p) = c[i];
+    }
+#else
+    (*p) -= sizeof(int);
+    *((int *)(*p)) = val;
+#endif
 
-        *--(*p) = 0x00;
-        len += 1;
+    for( len = sizeof(int) ; len > 0 ; len-- )
```

Comment:
I think this code is too complicated. It's easier to step through the integer byte by byte, starting with the lowest, and abort once we find a zero-byte. This can be done platform-independent bit-shifts.